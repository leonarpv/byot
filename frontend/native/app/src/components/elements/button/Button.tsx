import * as React from 'react';
import {Button as Btn, ButtonProps} from '@ui-kitten/components/ui/button/button.component';
import {StyleSheet, View} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import {baseTheme} from '../../../setup';

interface Props extends Omit<ButtonProps, 'status'> {
  color?: ButtonProps['status'] | 'gradient' | 'secondary';
}

interface State {
  touched: boolean;
}

const makeStyles = (props: Props, state: State) => {
  const shadow = {
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.32,
    shadowRadius: 5.46,

    elevation: 9,
    borderRadius: 32,
  };
  return StyleSheet.create({
    button: {
      ...((props.appearance || 'filled') === 'filled' && shadow),
    },
    gradientButton: {
      backgroundColor: '#00000000',
      borderWidth: 0,
    },
    buttonWrapper: {
      opacity: state.touched ? 0.8 : 1,
      ...((props.appearance || 'filled') === 'filled' && shadow),
    },
    gradient: {
      borderRadius: 32,
    },
  });
};

export const Button: React.FC<Props> = (props: Props) => {
  const [touched, setTouched] = React.useState<State['touched']>(false);
  const styles = makeStyles(props, {touched});
  const testID = 'element-button';

  return props.color !== 'gradient' ? (
    <Btn {...props} style={[props.style, styles.button]} accessibilityLabel={testID} testID={testID}>
      {props.children}
    </Btn>
  ) : (
    <View
      style={styles.buttonWrapper}
      onTouchStart={() => setTouched(true)}
      onTouchEnd={() => setTouched(false)}>
      <LinearGradient
        style={styles.gradient}
        start={{x: 0, y: 0}}
        end={{x: 1, y: 1}}
        colors={[baseTheme.colors.gradient.start.color, baseTheme.colors.gradient.end.color]}>
        <Btn
          {...props}
          style={[props.style, styles.gradientButton]}
          accessibilityLabel={testID}
          testID={testID}>
          {props.children}
        </Btn>
      </LinearGradient>
    </View>
  );
};
