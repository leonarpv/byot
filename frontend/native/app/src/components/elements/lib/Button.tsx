import * as React from 'react';
import {Button as Btn, ButtonPropTypes, Colors, Spacings, View} from 'react-native-ui-lib';
import {ButtonProps, StyleProp, StyleSheet, ViewStyle} from 'react-native';
import {Wave} from 'react-native-animated-spinkit';

type Props = Omit<ButtonProps, 'title'> &
  ButtonPropTypes & {
    variant?: 'primary' | 'secondary';
    rootStyle?: StyleProp<ViewStyle>;
    loading?: boolean;
  };

interface State {}

const makeStyles = (props: Props, state: State) =>
  StyleSheet.create({
    root: {
      borderRadius: 10,
      backgroundColor: Colors[props.variant || 'primary'],
      width: '100%',
      padding: 12,
      justifyContent: 'center',
      alignItems: 'center',
    },
    loaderBar: {
      borderRadius: 6,
    },
  });

export const Button: React.FC<Props> = (props: Props) => {
  const styles = makeStyles(props, {});
  const testID = props.testID || 'element-button';

  return props.loading ? (
    <View style={styles.root}>
      <Wave color={Colors.white} size={20} innerStyle={styles.loaderBar} />
    </View>
  ) : (
    <Btn testID={testID} accessibilityLabel={testID} {...props} />
  );
};
