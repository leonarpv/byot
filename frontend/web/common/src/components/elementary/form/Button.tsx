import React from 'react';
import baseTheme from '@byot-frontend/common/src/shared/theme/theme';
import {Button as MuiButton, ButtonProps, makeStyles, Theme} from '@material-ui/core';
import classNames from 'classnames';
import {StyleRules} from '@material-ui/styles';

interface Props extends Omit<ButtonProps, 'color'> {
  // todo variant: outlined gradient
  color: ButtonProps['color'] | 'gradient';
}

const styles = (theme: Theme): StyleRules<Props> => ({
  root: (props: Props) => ({
    ...(props.color === 'gradient' && {
      background: baseTheme.colors.gradient.css,
      color: baseTheme.colors.gradient.contrast,
    }),
    paddingTop: theme.spacing(1 / 2),
    paddingBottom: theme.spacing(1 / 2),
    textTransform: 'none',
    fontWeight: 400,
  }),
  outlined: {
    paddingTop: theme.spacing(1 / 2),
    paddingBottom: theme.spacing(1 / 2),
  },
});
const useStyles = makeStyles(styles);

export const Button: React.FC<Props> = (props: Props) => {
  const styles = useStyles(props);
  const color = props.color === 'gradient' ? 'inherit' : props.color;
  const variant = props.color === 'gradient' ? 'contained' : props.variant;
  return (
    <MuiButton
      {...props}
      color={color}
      variant={variant}
      classes={{
        ...(props.classes || {}),
        root: classNames(styles.root, props.classes?.root),
        outlined: classNames(styles.outlined, props.classes?.outlined),
      }}
    />
  );
};
