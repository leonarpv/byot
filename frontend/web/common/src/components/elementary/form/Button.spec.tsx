import * as React from 'react';
import {Button} from './Button';
import {Button as MuiButton, createMuiTheme, Theme} from '@material-ui/core';
import ScaleLoader from 'react-spinners/ScaleLoader';
import {Shallow} from '../../../test/CreateShallow';
import {createShallow} from '@material-ui/core/test-utils';

describe('<Button/>', () => {
  let shallow: Shallow;
  let theme: Theme;

  beforeEach(() => {
    theme = createMuiTheme();
    shallow = createShallow({});
  });

  it('should display button with text', () => {
    const wrapper = shallow(<Button>Text</Button>);
    const span = wrapper.find('[data-testid="common-elementary-button-children"]');
    expect(span.length).toEqual(1);
    expect(span.children().text()).toEqual('Text');
    const loader = wrapper.find('[data-testid="common-elementary-button-loader"]');
    expect(loader.length).toEqual(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('should display button with loader', () => {
    const wrapper = shallow(<Button loading>Text</Button>);
    const span = wrapper.find('[data-testid="common-elementary-button-children"]');
    expect(span.length).toEqual(1);
    expect(span.children().text()).toEqual('Text');
    const loader = wrapper.find('[data-testid="common-elementary-button-loader"]');
    expect(loader.length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should display primary button with contrast loader', () => {
    const color = 'primary';
    const wrapper = shallow(
      <Button loading color={color}>
        Text
      </Button>
    );
    const span = wrapper.find('[data-testid="common-elementary-button-children"]');
    expect(span.length).toEqual(1);
    expect(span.children().text()).toEqual('Text');
    const loader = wrapper.find('[data-testid="common-elementary-button-loader"]');
    expect(loader.length).toEqual(1);
    const muiButton = wrapper.find(MuiButton);
    expect(muiButton.prop('color')).toEqual(color);
    const scaleLoader = wrapper.find(ScaleLoader);
    expect(scaleLoader.prop('color')).toEqual(theme.palette.primary.contrastText);
    expect(wrapper).toMatchSnapshot();
  });

  it('should display primary button with primary loader', () => {
    const color = 'primary';
    const wrapper = shallow(
      <Button loading variant="outlined" color={color}>
        Text
      </Button>
    );
    const span = wrapper.find('[data-testid="common-elementary-button-children"]');
    expect(span.length).toEqual(1);
    expect(span.children().text()).toEqual('Text');
    const loader = wrapper.find('[data-testid="common-elementary-button-loader"]');
    expect(loader.length).toEqual(1);
    const muiButton = wrapper.find(MuiButton);
    expect(muiButton.prop('color')).toEqual(color);
    const scaleLoader = wrapper.find(ScaleLoader);
    expect(scaleLoader.prop('color')).toEqual(theme.palette[color].main);
    expect(wrapper).toMatchSnapshot();
  });
});
