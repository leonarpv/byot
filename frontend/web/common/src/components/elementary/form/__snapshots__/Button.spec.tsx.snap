// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should display button with loader 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-root-6 makeStyles-root-9 MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="makeStyles-inner-7 makeStyles-inner-10"
            data-testid="common-elementary-button-children"
            style="opacity: 0; visibility: hidden;"
          >
            Text
          </span>
          <div
            class="makeStyles-loader-8"
            data-testid="common-elementary-button-loader"
            style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          >
            <div
              class="css-0"
            >
              <div
                class="css-1ab95gp"
              />
              <div
                class="css-tm4xur"
              />
              <div
                class="css-1epl0i6"
              />
              <div
                class="css-1fj2adh"
              />
              <div
                class="css-1vivfwt"
              />
            </div>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-root-6 makeStyles-root-9 MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="makeStyles-inner-7 makeStyles-inner-10"
          data-testid="common-elementary-button-children"
          style="opacity: 0; visibility: hidden;"
        >
          Text
        </span>
        <div
          class="makeStyles-loader-8"
          data-testid="common-elementary-button-loader"
          style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          <div
            class="css-0"
          >
            <div
              class="css-1ab95gp"
            />
            <div
              class="css-tm4xur"
            />
            <div
              class="css-1epl0i6"
            />
            <div
              class="css-1fj2adh"
            />
            <div
              class="css-1vivfwt"
            />
          </div>
        </div>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Button/> should display button with text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-root-1 makeStyles-root-4 MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="makeStyles-inner-2 makeStyles-inner-5"
            data-testid="common-elementary-button-children"
            style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          >
            Text
          </span>
          <div
            class="makeStyles-loader-3"
            data-testid="common-elementary-button-loader"
            style="opacity: 0; visibility: hidden;"
          >
            <div
              class="css-0"
            >
              <div
                class="css-1ab95gp"
              />
              <div
                class="css-tm4xur"
              />
              <div
                class="css-1epl0i6"
              />
              <div
                class="css-1fj2adh"
              />
              <div
                class="css-1vivfwt"
              />
            </div>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-root-1 makeStyles-root-4 MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="makeStyles-inner-2 makeStyles-inner-5"
          data-testid="common-elementary-button-children"
          style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          Text
        </span>
        <div
          class="makeStyles-loader-3"
          data-testid="common-elementary-button-loader"
          style="opacity: 0; visibility: hidden;"
        >
          <div
            class="css-0"
          >
            <div
              class="css-1ab95gp"
            />
            <div
              class="css-tm4xur"
            />
            <div
              class="css-1epl0i6"
            />
            <div
              class="css-1fj2adh"
            />
            <div
              class="css-1vivfwt"
            />
          </div>
        </div>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
